### PR TITLE
add include/exclude_paths to the auth plugins

### PR DESCRIPTION
**Pull request by ammmze**
I believe this may satisfy #853.

We have a use case where we want some endpoints within our api to be able to be accessible anonymously. Each of our api's hosts its own swagger ui with api-docs which we want to be accessible anonymously while securing the other endpoints with oauth2. The most flexible solution I thought of was having `include_paths` and `exclude_paths` arrays in the plugin config. Using these you can get pretty good control over which endpoints require authentication and which do not. This also opens the possibility of having multiple auth plugins on a single api for different endpoints (example basic-auth on the swagger ui and oauth2 on all else).

Each element in the include/exclude path arrays is a regex that is applied to the request url (after stripping out the path prefix). 

Also if you do hit an endpoint which is excluded and you do have the necessary auth header(s), it will still try authentication. It will only skip the auth stuff if the auth headers are not present.

Note: I can't seem to figure out the deal with the travis ci build :( It looks like it fails sending the coveralls report? I was hoping if I created a coveralls account and added my fork that it would work, but it didn't seem to help.

Just getting the following error (even if i build the latest `next` branch):
```
Upload error: Couldn't find a repository matching this job.
Upload error: Couldn't find a repository matching this job.
Upload error: Couldn't find a repository matching this job.
```

### CODE REVIEW

**Comment by ammmze** [comment]
Would we prefer moving that `is_path_included` method to a more global location? I wasn't sure how isolated you want the plugins to be.

**Comment by thibaultCha** [comment]
Hi,

As much as we appreciate the contribution, we currently have plans to allow unauthenticated calls, and those are part of a bigger plan to allow more flexibility with APIs and Plugins. We would like all plugins (not only authentication ones) to have such features as include/exclude path, and the model refactoring needed requires a lot of reflexion to satisfy all our use cases (which is to be a less opinionated as possible). This is a process we are currently undergoing.

When such a feature/model refactoring will be implemented, it should also, as discusses in #853, introduce the notion of "anonymous consumer", a consumer that has less powers than an authenticated one, and such a concept impacts both the core and the sibling plugins.

In short, this is a much deeper problem that does not only impact the authentication plugins. We are happy that you could find a solution for your use case by tweaking Kong to your needs, but to benefit to the entire community, a lot more work will be required.

PS: If you are to use this, I would indeed recommend grouping those functions together, and avoid duplicating that much code. This is a very good hint as to why this issue needs to be addressed at a deeper level.

PPS: the coveralls.io error you are seeing suddenly popped up yesterday and is not only related to your branch/PR but also affects branches of this repository. We haven't figured it out yet.

**Comment by thibaultCha** [comment]
(Leaving this opened for discussion, feel free to share thoughts!)

**Comment by thefosk** [comment]
Related to #505.
